addons/project/models/project.py

```diff
@@ -1796,9 +1796,14 @@ def write(self, vals):
             if 'kanban_state' not in vals:
                 vals['kanban_state'] = 'normal'
         # user_ids change: update date_assign
-        if vals.get('user_ids') and 'date_assign' not in vals:
-            vals['date_assign'] = now
-
+        if not vals.get('user_ids', []) and 'user_ids' in vals:
+            vals['date_assign'] = False
+        elif 'user_ids' in vals:
+            new_user_ids = self._fields['user_ids'].convert_to_cache(vals['user_ids'], self.env['project.task'])
+            if len(new_user_ids) == 0:
+                vals['date_assign'] = False
+            elif 'date_assign' not in vals:
+                vals['date_assign'] = now
```

Comment:
We could only test once that 'user_ids' is in vals dict.
I don't think the convert_to_cache will work in case of an unlink.
It will be best to test is after the call to super.
You can keep the first part prior to the call so super.